sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala

```diff
@@ -106,10 +106,8 @@ abstract class CompactibleFileStreamLog[T <: AnyRef : ClassTag](
     interval
   }
 
-  /**
-   * Filter out the obsolete logs.
-   */
-  def compactLogs(logs: Seq[T]): Seq[T]
+  /** Determine whether the log should be retained or not. */
+  def shouldRetain(log: T): Boolean
```

Comment:
I just retained the functionality of exclusion, as we still have a chance to apply retention which is applied per entry.

Reply:
`compactLogs` requires to pass "all entries" which should be loaded into memory, right?

Reply:
The UT I had to remove leverages this behavior - it's only possible when we collect all entries, as it assumes DELETE_ACTION can come after ADD_ACTION "in any further batches" and compactLogs should be able to filter out. This also requires us to materialize all entries.

Reply:
I take some time looking into the detail. Seems like we can use `Stream` here, which can delay the evaluation for each file status. Then we can both keep the API and have the improvement together. But for using `Stream[Int]` here, we'll have a restriction for the implementation of `compactLogs`, which also needs to return `Stream[Int]`, otherwise, the improvement of streamline will not take effect.

Reply:
Is there a specific reason to avoid changing the method which is private API? I'd say it'd be less clear to provide the intention on downstream (technically we shouldn't concern with downstream though) - if you don't change the method signature but just change the semantic, no one would know about the change and it will be no-op for downstream.

I'll consider it if you have strong reason to do so, otherwise let's hear other's voice as well. cc. @zsxwing

Reply:
Maybe not strong enough. IMO if we are able to avoid breaking the third-party libs, we'd better do so. Yep, let's hear more voice.
Great thanks for your patience and reply. It's the last concern left, other changes for this PR LGTM. I do think we need this improvement!